File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.cpp
class Solution {
public:
    int numberOfArithmeticSlices(vector<int>& nums) {
        int n = nums.size();
        long long res = 0;
        vector<map<long long, int>>dp(n);
        for(int i = 1; i<n; i++){
            for(int j = 0; j<i; j++){
                long long d = (long long)nums[i] - (long long) nums[j];
                if(d<= INT_MIN || d >= INT_MAX){
                    continue;
                }
                int s = 0;
                if(dp[j].find(d) != dp[j].end()){
                    s = dp[j][d];
                }
                dp[i][d] += s + 1;
                res += s;
            }
        }
        return (int)res;
    }
};

File: 0446-arithmetic-slices-ii-subsequence/0446-arithmetic-slices-ii-subsequence.py
class Solution:
    def numberOfArithmeticSlices(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [defaultdict(int) for _ in range(n)]
        res = 0
        for i in range(1,n):
            for j in range(i):
                d = nums[i]-nums[j]
                s = 0
                if d in dp[j]:
                    s =  dp[j][d]
                dp[i][d] += s +1
                res += s
        return res